File: app/routes/motor.py
from flask import Blueprint, request, jsonify

from src.state import get_state_collector
from app.services import get_control_service

motor_bp = Blueprint("motor", __name__, url_prefix="/api/motor")


@motor_bp.route("/status")
def motor_status():
    return jsonify(get_state_collector().get_state())


@motor_bp.route("/direct", methods=["GET"])
def motor_direct():
    left = int(float(request.args.get('left', 0)))
    right = int(float(request.args.get('right', 0)))
    duration = float(request.args.get('duration', 0))

    try:
        result = get_control_service().run_motor(left, right, duration)
        status = get_state_collector().get_state()
        result["left_speed"] = status.get("left_speed", 0)
        result["right_speed"] = status.get("right_speed", 0)
        return jsonify(result)
    except Exception as e:
        return jsonify({"error": str(e)}), 500


@motor_bp.route("/raw_command", methods=["GET"])
def raw_command():
    cmd = request.args.get('cmd', '')
    return jsonify(get_control_service().send_raw_command(cmd))

File: dora/web-server/src/routes/motor.rs
//! 电机 HTTP 路由 —— 匹配前端 api.ts 中的 motor.* 调用
//!
//! 对应 `app/routes/motor.py`。

use std::sync::Arc;

use axum::{
    extract::{Query, State},
    Json, Router,
};
use axum::routing::get;
use serde::Deserialize;

use crate::AppState;

pub fn router() -> Router<Arc<AppState>> {
    Router::new()
        .route("/api/motor/status", get(status))
        .route("/api/motor/direct", get(direct))
        .route("/api/motor/raw_command", get(raw_command))
}

#[derive(Deserialize, Default)]
struct DirectParams {
    left: Option<i32>,
    right: Option<i32>,
    duration: Option<u32>,
}

async fn direct(
    State(s): State<Arc<AppState>>,
    Query(p): Query<DirectParams>,
) -> Json<serde_json::Value> {
    s.motor.direct(p.left.unwrap_or(0), p.right.unwrap_or(0), p.duration.unwrap_or(0)).await;
    Json(serde_json::json!({"ok": true}))
}

async fn status(State(s): State<Arc<AppState>>) -> Json<serde_json::Value> {
    let s = s.motor.status();
    Json(serde_json::json!({
        "left_speed":  s.left_speed,    // m/s
        "right_speed": s.right_speed,   // m/s
    }))
}

async fn raw_command(
    State(s): State<Arc<AppState>>,
    Query(p): Query<std::collections::HashMap<String, String>>,
) -> Json<serde_json::Value> {
    let cmd = p.get("cmd").map(String::as_str).unwrap_or("");
    let result = s.motor.raw_command(cmd).await;
    Json(result)
}
